django/db/models/query_utils.py

```diff
@@ -340,38 +340,37 @@ def _unregister_instance_lookup(self, lookup, lookup_name=None):
     _unregister_class_lookup = classmethod(_unregister_class_lookup)
 
 
-def select_related_descend(field, restricted, requested, select_mask, reverse=False):
+def select_related_descend(field, restricted, requested, select_mask):
     """
-    Return True if this field should be used to descend deeper for
-    select_related() purposes. Used by both the query construction code
-    (compiler.get_related_selections()) and the model instance creation code
-    (compiler.klass_info).
+    Return wether `field` should be used to descend deeper for
+    `select_related()` purposes.
 
     Arguments:
-     * field - the field to be checked
-     * restricted - a boolean field, indicating if the field list has been
-       manually restricted using a requested clause)
-     * requested - The select_related() dictionary.
-     * select_mask - the dictionary of selected fields.
-     * reverse - boolean, True if we are checking a reverse select related
+     * `field` - the field to be checked. Can be either a `Field` or
+       `ForeignObjectRel` instance.
+     * `restricted` - a boolean field, indicating if the field list has been
+       manually restricted using a select_related() clause)
+     * `requested` - the select_related() dictionary.
+     * `select_mask`
     """
+    # Only relationships can be descended.
     if not field.remote_field:
         return False
-    if field.remote_field.parent_link and not reverse:
+    # Forward MTI parent links should not be explicitly descended as they are
+    # always JOIN'ed against unless excluded by `select_mask`.
+    if getattr(field.remote_field, "parent_link", False):
         return False
```

Comment:
In other words, for the following models

```python
class Parent(models.Model):
   pass

class Child(Parent):
    pass
```

There's no need to do anything special for `Child.parent_ptr` (the implicit `parent_link` primary key) as it will be always be implicitly followed when doing `Child.objects.all()`.